daemon/policy.go

```diff
@@ -276,7 +277,41 @@ func (d *Daemon) PolicyDelete(labels labels.LabelArray) *apierror.APIError {
 		return apierror.New(DeletePolicyNotFoundCode, "policy not found")
 	}
 
-	d.TriggerPolicyUpdates([]policy.NumericIdentity{})
+	go func() {
+		wg := d.TriggerPolicyUpdates([]policy.NumericIdentity{})
+		// Wait for all policies to be updated so that
+		// we can grab a fresh map of consumables.
+		wg.Wait()
+
+		consumables := d.consumableCache.GetConsumables()
+
+		endpointmanager.Mutex.RLock()
+		for _, ep := range endpointmanager.Endpoints {
+			ep.Mutex.RLock()
+			// If the policy is not being enforced then keep the CT
+			// entries
+			// FIXME: Do we really need to check the daemon.PolicyEnforcement()?
+			if ep.SecLabel == nil ||
+				!ep.Opts.IsEnabled(endpoint.OptionPolicy) ||
+				d.PolicyEnforcement() == endpoint.NeverEnforce {
+				ep.Mutex.RUnlock()
+				continue
+			}
+			epSecID := ep.SecLabel.ID
+
+			ep.Mutex.RUnlock()
+
+			idsToKeep := map[uint32]bool{}
+			if consumers, ok := consumables[epSecID]; ok {
+				for _, consumer := range consumers {
+					idsToKeep[consumer.Uint32()] = true
+				}
+			}
+			log.Debugf("Keeping entries of EP %d: %+v", ep.ID, idsToKeep)
+			endpointmanager.KeepCTEntryOf(!d.conf.IPv4Disabled, ep, idsToKeep)
```

Comment:
Let me get this straight:

1. policy deleted
    1. waiting for consumables map to be regenerated
    2. retrieve a copy of consumables and generate the `IDsToKeep`
    3. start cleaning up all CT entries except `IDsToKeep`

2. a new (allowed by the current/new policy) consumer starts talking with the consumable after 1.ii

3. a new (denied by the new policy) consumer starts talking with the consumable after 1.ii

- [x] the endpoint created in 3 will be denied by the policy so there won't even be a CT entry.

- [ ] the endpoint created in 2 will be allowed by the policy but the CT entry will be removed because it's not in `IDsToKeep`

Ok, what do you suggest? I was thinking on holding the endpoint lock until all CT entries are cleaned.

This guarantees the endpoint won't get its `allowed-consumers` updated until its CT entries are cleaned up. But this also means I will need to move the `d.consumableCache.GetConsumables()` inside the for loop and get the `IDsToKeep` for each endpoint causing different endpoints (with the same security ID) to get different `IDsToKeep` map.